### PR TITLE
New Workshop: Polish

### PR DESCRIPTION
**Pull request by SallyMcGrath**
an afternoon to spend together making things look great
I've targeted the main obvious easy wins that our trainees usually fall down on
I've not explicitly set aside time to write a readme just expected it to exist (even though it probably doesn't
line changes please, if you can!

### CODE REVIEW

**Review by 40thieves — COMMENTED** [review]
Not sure if you wanted a review yet, but I really like this so far 👍🏻

I've left a couple of very minor suggestions

File: polish/readme.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+- [ ] Colours are consistent, drawn from a [limited, designed palette](https://coolors.co/palettes/trending)
+- [ ] There are no contrast issues (Lighthouse should have caught this, but run the [Accessibility Insights](https://accessibilityinsights.io/) to check again).
+- [ ] Spacing is consistent, drawn from a [limited, designed scale](https://uxplanet.org/spacing-guide-for-designers-5bd140afe52a)
+
+It may be helpful to know the [user agent](https://chromium.googlesource.com/chromium/blink/+/master/Source/core/css/html.css) type scale is roughly 10px, 12px, 16px, 18px, 24px, 32px.
+
+Get out of your own way a bit here. If you are only looking for Cloud roles or something without any visual interface development, then use [Tailwind](https://tailwindcss.com/), which has all this pre-set for you and will look basically fine.
+
+What you must do is make your UI look _intentional_. It's fine to be experimental, it's fine to be basic, it's fine to use a design system, but it's not fine to have a UI that looks like you're not in control of it.
+
```
[the remark this answers is not in the document]
Comment (SallyMcGrath) [reply polish/readme.md:68]: That's a good idea. I wonder if we can connect it to the components workshop as well  - they have this prep https://github.com/CodeYourFuture/Module-HTML-CSS/issues/52 for that one. This workshop tries NOT to assume when it is happening, other than they have a website already, so, hm, maybe I'm talking myself out of connecting...it's a good idea though. 

Incidentally, when I wrote this, GitHub CoPilot tried to autocomplete like this:

What you must do is make your UI look intentional. It's fine to be basic, it's fine to use a design system, but it's not `fine to have a UI that looks like it was thrown together by a child. If you are not a designer, then use a design system. If you are a designer, then use your design skills. If you are not a designer and you don't use a design system, then you are a bad developer.`

😂

File: polish/readme.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
+# Polish
+
+In this workshop we will take our existing projects and polish them to a higher standard. We will use Lighthouse to evaluate our work and make improvements. We will apply some basic design principles to bring a professional look to our work. (We can talk about polishing code in another workshop.)
```
[the remark this answers is not in the document]
Comment (SallyMcGrath) [reply polish/readme.md:3]: Always want reviews! Thank you!!

File: polish/readme.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
+# Polish
```
Comment (SallyMcGrath) [inline polish/readme.md:1]:
```suggestion
# UI Polish
```

File: polish/readme.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+### ⚠️ Note: Size your images correctly
+
+Lighthouse should now pick up images with popped aspect ratios. If you have any images that are not sized correctly, fix them. Compress your images and deliver appropriate sizes to appropriate devices. Janky looking pictures make you look amateurish. Lazy load any images [below the fold](https://web.dev/lazy-loading-best-practices/). You can do this with an attribute or a library.
+
+### 🎨 2. Fonts, colours, and spacing
+
+#### ✅ Checklist
+
+- [ ] Fonts are consistent, drawn from a [limited, designed scale](https://typescale.com/)
```
[the remark this answers is not in the document]
Comment (SallyMcGrath) [reply polish/readme.md:58]: That is a great link!

Comment (SallyMcGrath) [inline polish/readme.md:58]:
```suggestion
- [ ] Fonts are consistent, drawn from a [limited, designed scale](https://typescale.com/) (Read more about type scales)[https://designcode.io/typographic-scales]
```

File: polish/readme.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+- [ ] Gutters are consistent, drawn from a [limited, designed scale](https://type-scale.com/)
+- [ ] Components are laid on a grid, with consistent gutters
+
+<!--{{note type="activity" title="Line up your components 🍅"}}>-->
+
+## Line up your components
+
+1. Take a piece of paper and sketch your page layout as it is now
+1. Take a new piece of paper and sketch your page layout as you want it to be
+1. Draw a grid over your sketch (5m)
```
Comment (SallyMcGrath) [inline polish/readme.md:95]:
```suggestion
1. Draw a grid over your sketch 
```

File: polish/readme.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (40thieves) [reply polish/readme.md:68]: I think having a thematic connection to that exercise is a good idea, if not a direct one.

I like the list of sites there - a good mix of use cases - so I wonder if the exercise is just to look at those 3 sites to identify what makes them look consistent? (I think that's mostly what we're getting at here?)

ChatGPT came up with this suggestion:

> Review the main navigation menu on the website. Consider whether the menu items and their placement appear intentional and aligned with the website's purpose and content.
> Explore different sections or pages within the website (e.g., articles, video pages, financial advice sections). Assess whether design elements (buttons, typography, color schemes) appear to serve a clear purpose and support the content or functionality.
> Take note of any branding elements, such as logos or color schemes, and evaluate whether they reflect the website's intended identity and message.

I'm not in love with that response - points 1 & 3 feel more like branding than UI design to me. It's a starting point? I'll have a think about whether I can phrase this exercises a bit better - happy to merge without this.

**Comment by 40thieves** [comment]
Playing around with ChatGPT to come up with some ideas for the exercise I suggested, I asked it for design principles for junior web developers - I think this response was pretty good:

> 1. **Visual Hierarchy:** Arrange elements on the page to guide users' attention. Use larger fonts, bold text, or contrasting colors to emphasize important content or actions.
> 2. **Grid Layout:** Employ a grid system to create a structured and organized layout. Grids help maintain alignment and consistency in the design.
> 3. **Typography:** Pay attention to font selection, size, line spacing, and line length. Ensure text is readable, and choose fonts that match the overall tone and style of the website.
> 4. **Color Scheme:** Select a harmonious color palette that reflects the brand and evokes the desired emotions. Consider color psychology to influence user behavior and perception.
> 5. **Whitespace:** Use whitespace effectively to separate and group content, making it easier to digest. Avoid overcrowding elements on the page.
> 6. **Buttons and Calls to Action (CTAs):** Make buttons and CTAs stand out with contrasting colors and clear labels. Ensure they are large enough to be easily clickable.
> 7. **Icons and Imagery:** Use icons and images to convey information quickly. Ensure that they are relevant and add value to the content.
> 8. **Consistent Styling:** Maintain consistent styling for elements like buttons, forms, headings, and links throughout the website. This helps users understand how different elements function.
> 9. **Responsive Design:** Ensure that your UI adapts gracefully to various screen sizes and orientations. Test your design on different devices to ensure usability.
> 10. **Navigation:** Create a user-friendly and intuitive navigation menu. Use clear labels and consider using navigation patterns like dropdown menus or hamburger menus for mobile devices.
> 11. **Forms and Input Fields:** Design forms that are easy to complete by using clear labels, appropriate input types, and error messages. Consider grouping related fields together.
> 12. **Feedback and Animation:** Provide feedback to users when they interact with elements (e.g., button clicks, form submissions) through animations or microinteractions. Keep animations subtle and purposeful.
> 13. **Loading Screens:** Design loading screens or spinners to keep users informed when content is loading, preventing frustration.
> 14. **Consistency Across Pages:** Ensure a consistent look and feel across all pages of the website to maintain a coherent user experience.
> 15. **Accessibility:** Even in UI design, consider accessibility by using text alternatives for images, providing keyboard navigation options, and ensuring good color contrast for readability.
> 16. **User Testing:** Whenever possible, conduct user testing to gather feedback and make improvements based on actual user experiences.
> 17. **Prototyping:** Create interactive prototypes to visualize how the UI elements will function and gather feedback from stakeholders early in the design process.
> 18. **Feedback Loops:** Be open to feedback and iterate on your design. Junior designers should continually learn and improve their skills.
> 19. **Design Patterns:** Familiarize yourself with common UI design patterns and best practices, such as card layouts, hero banners, and content carousels.
> 20. **Simplicity:** Keep the UI design clean and uncluttered. Remove any elements or features that do not add value to the user experience.

I think we've covered most of the points that I think are relevant for our trainees, I wanted to pull out visual hierarchy in particular as something we could look to add. I think that's one area where I could sense a design was bad, but couldn't figure out why.

Maybe we could add a point to the checklist along these lines

> There is [visual hierarchy](https://www.nngroup.com/articles/visual-hierarchy-ux-definition/)

Then add a section something like:

> ### Squint test  👀 
> 1. Try squinting at your page, taking your glasses off or stepping far away from the screen
> 2. Is your eye guided through the page in the order that you want? Is there anything standing out that isn't important?
> 3. Try tweaking colours, font sizes and weights or spacing to emphasis or de-emphasise parts of the page
> 4. Repeat the squint test after you've made the changes. Is there a visual order? Keep iterating until you're happy with the flow of the page

---

This thought (sorry, hopefully this chain of thought is followable) also had me remembering a old video series from one of the folks behind Tailwind https://www.youtube.com/playlist?list=PLDVpvW8ghDr9tasku_YvuTy_l0xPUtOHE The idea was essentially for their designer to reproduce pages in his design tool and then "refactor" them into nicer looking pages, with commentary on why he was making those choices.

I think my design taste has evolved a bit since then, but I thought at the time they were really helpful in explaining design principles with concrete examples. Perhaps you could review, and if you're happy then we could set as coursework?

**Comment by SallyMcGrath** [comment]
So I think instead of us giving them all this info, maybe having them ask ChatGPT and do the revisions? 

> I think my design taste has evolved a bit since then, but I thought at the time they were really helpful in explaining design principles with concrete examples. Perhaps you could review, and if you're happy then we could set as coursework?

**Pull request merged**